### ISSUE TITLE
run code and send ss

### ISSUE THREAD

**Issue by kkiritok** [kkiritok/Quantam-based-optimization#19]
import torch
import torch.nn as nn
import torchvision
import torchvision.transforms as transforms
from torch.utils.data import DataLoader
from torchvision.models import resnet18

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print("Using device:", device)

# ======================
# Dataset (MNIST)
# ======================
train_transform = transforms.Compose([
    transforms.Resize(224),   # ResNet input size
    transforms.Grayscale(num_output_channels=1),
    transforms.ToTensor(),
    transforms.Normalize(
        mean=(0.1307,),   # MNIST mean
        std=(0.3081,)     # MNIST std
    )
])

test_transform = transforms.Compose([
    transforms.Resize(224),
    transforms.Grayscale(num_output_channels=1),
    transforms.ToTensor(),
    transforms.Normalize(
        mean=(0.1307,),
        std=(0.3081,)
    )
])

trainset = torchvision.datasets.MNIST(
    "./data", train=True, download=True, transform=train_transform
)

testset = torchvision.datasets.MNIST(
    "./data", train=False, download=True, transform=test_transform
)

trainloader = DataLoader(trainset, batch_size=128, shuffle=True, num_workers=2)
testloader = DataLoader(testset, batch_size=128, shuffle=False, num_workers=2)

# ======================
# Model (FREEZE LAYER 1–4)
# ======================
model = resnet18(pretrained=True)

# 🔥 IMPORTANT: Change input channel (3 → 1)
model.conv1 = nn.Conv2d(
    1, 64, kernel_size=7, stride=2, padding=3, bias=False
)

# Freeze all layers
for param in model.parameters():
    param.requires_grad = False

# Replace final FC layer (100 → 10)
model.fc = nn.Linear(512, 10)

# Unfreeze only FC layer
for param in model.fc.parameters():
    param.requires_grad = True

model = model.to(device)

criterion = nn.CrossEntropyLoss()

optimizer = torch.optim.AdamW(
    model.fc.parameters(),   # only train FC
    lr=3e-4,
    weight_decay=1e-4
)

# ======================
# Accuracy function
# ======================
def accuracy(model, loader):
    model.eval()
    correct = 0
    total = 0
    with torch.no_grad():
        for x, y in loader:
            x, y = x.to(device), y.to(device)
            logits = model(x)
            preds = logits.argmax(dim=1)
            correct += (preds == y).sum().item()
            total += y.size(0)
    return 100 * correct / total

# ======================
# Training Loop
# ======================
EPOCHS = 5

print("\n🚀 TRAINING MNIST (FC ONLY, ALL OTHER LAYERS FROZEN)")
for epoch in range(EPOCHS):
    model.train()
    loss_sum = 0.0

    for x, y in trainloader:
        x, y = x.to(device), y.to(device)

        optimizer.zero_grad()
        logits = model(x)
        loss = criterion(logits, y)
        loss.backward()
        optimizer.step()

        loss_sum += loss.item()

    train_acc = accuracy(model, trainloader)
    test_acc = accuracy(model, testloader)

    print(
        f"Epoch {epoch+1}/{EPOCHS} | "
        f"Loss {loss_sum/len(trainloader):.4f} | "
        f"Train Acc {train_acc:.2f}% | "
        f"Test Acc {test_acc:.2f}%"
    )
